dnsmasq-metrics/pkg/server/dnsprobe_test.go

```diff
@@ -0,0 +1,137 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package server
+
+import (
+	"fmt"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/miekg/dns"
+
+	"k8s.io/contrib/dnsmasq-metrics/pkg/test"
+)
+
+func makeOptions(label string, addr string, port int) *Options {
+	return &Options{
+		PrometheusNamespace: "test",
+		Probes: []DNSProbeOption{
+			{
+				Label:    label,
+				Server:   fmt.Sprintf("%v:%v", addr, port),
+				Name:     "test.local.",
+				Interval: 1 * time.Millisecond,
+			},
+		},
+	}
+}
+
+func TestProbeOk(t *testing.T) {
+	server := &test.Server{}
+	addr, port := server.Init(t)
+	go server.Run(okResponseCallback)
+	<-server.StartChan
+
+	options := makeOptions("ok", addr, port)
+	probe := &dnsProbe{DNSProbeOption: options.Probes[0]}
+	probe.Start(options)
+
+	// This is ugly and flaky but we wait for some probes to go through by sleeping.
+	time.Sleep(250 * time.Millisecond)
```

Comment:
Suggest to use a mock clock to prevent flakiness. As we did in https://github.com/kubernetes/contrib/blob/master/exec-healthz/cmd/exechealthz/exechealthz_test.go#L55-L84.